Respect SET/RESET SESSION queries

saurfang opened the pull request:
According #32, parameters can already be set via `dbConnect(..., parameters = list(...)`.

The Python implementation of Presto API appears to respect [`SET/RESET SESSION ...`](https://prestodb.io/docs/current/sql/set-session.html) query commands directly. This PR does the same for RPResto according to [HTTP Protocol](https://github.com/prestodb/presto/wiki/HTTP-Protocol) specification.

One caveat is this is only possible via `dbGetQuery` due to the API behavior, where parameters are managed by the client instead of the server.

I had to add a new RCpp function to ensure parameters are updated in-place instead of being copied upon modification inside the function.
Furthermore, `PrestoResult` doesn't have access to the connection object and therefore gains a new function slot to do the update.

Test cases are added and tested with a local cluster.

Updated to Roxygen 6.0.1

onurfiliz commented on RPresto.Rproj:
```diff
@@ -5,8 +5,13 @@ SaveWorkspace: No
 AlwaysSaveHistory: Default
 
 EnableCodeIndexing: Yes
+UseSpacesForTab: Yes
```
Could you remove the RPresto.Rproj changes? I'd prefer to keep the context of the PR limited to session parameters.

saurfang replied:
done

onurfiliz commented on tests/testthat/test-session.property.R:
```diff
@@ -0,0 +1,54 @@
+
+query_session_property <- function(conn, property) {
+  properties <- dbGetQuery(conn, 'SHOW SESSION')
+  properties[properties$Name == property, ]$Value
```
Please use list[['item']] style of indexing for accessing list items.

You can also do:

    return(properties[properties[['Name']] == property, 'Value', drop=TRUE])

saurfang replied:
done

onurfiliz commented on tests/testthat/test-session.property.R:
```diff
@@ -0,0 +1,54 @@
+
+  # verify default value is applied
+  test_property <- 'optimize_hash_generation'
+  expect_equal(query_session_property(conn, test_property), 'true')
```
Even though it's a rather safe assumption, this default value would not be guaranteed. A better option would be to take the default value and invert it in the following tests.

saurfang replied:
done

onurfiliz commented on R/dbFetch.R:
```diff
@@ -75,7 +99,10 @@ NULL
     chunk.count <- chunk.count + 1
   }
-  if (length(rv) == 1) {
+  if (length(rv) == 0) {
```
I'd prefer if we didn't do this. If the server does not return anything even when we have fetched all chunks, then it's either a server bug or we are not handling errors properly at some other spot.

onurfiliz commented on R/dbFetch.R:
```diff
@@ -61,6 +61,30 @@ NULL
+      switch(
+        content[['updateType']],
+        'SET SESSION' = {
+          properties <- get.response$headers[['x-presto-set-session']]
```
Can you use `httr::headers()` instead?

onurfiliz commented on R/dbSendQuery.R:
```diff
@@ -49,7 +49,10 @@ NULL
-        cursor=cursor
+        cursor=cursor,
+        update.conn.parameter=function(property, value) {
+          set_connection_parameter(conn, property, value)
```
The RCpp route feels rather hacky. My suggestion would be to use a reference class (e.g. named PrestoSession) for storing the parameters in the connection - similar to what we are doing with the PrestoCursor. Then we can carry the instance of that class around in the PrestoResult to update as necessary.
We could have a method in PrestoConnection that returns the required list by reading the values from the reference class instance.

saurfang commented:
Thank you for the review. I addressed all your comments and migrated to use a Reference class instead. However, I had to drop `parameters` slot from PrestoConnection to avoid storing potential conflict parameters in two places. The `dbConnect` interface is kept the same so unless users have been access `conn@parameters` directly, this should be fine. PTAL.

onurfiliz commented on R/PrestoSession.R:
```diff
@@ -0,0 +1,40 @@
+    parameters=function() {
+      return(.parameters)
+    },
+    parameterStr=function() {
```
nit: I'd call this `parameterString`.

saurfang commented:
Sure thing. I also squashed the commits and reran the tests. It seems fine from my end.

FYI some tests are failing for me because `'src_sql' is not an exported object from 'namespace:dplyr'` in `dplyr` release candidate.

onurfiliz commented:
Hmm, the `src_sql` errors should be fixed by #76, you may have needed to rebase? In any case, I'll merge as is and verify that the tests pass. Thanks again.
